### PR TITLE
Add function to solve quadratic equations

### LINKED ISSUE #397: Add functionality to solve quadratic equation

### PR DESCRIPTION
**Pull request by hamirmahal**
closes #397.

### CODE REVIEW

File: numbat/tests/interpreter.rs [hunk 1]
```diff
@@ -227,6 +227,18 @@ fn test_function_inverses() {
 
 #[test]
 fn test_math() {
+    expect_output("quadratic_equation(1, 1, -1)", "(0.618034, -1.61803)");
+    expect_output("quadratic_equation(9.0, -126.0, 441)", "(7, 7)");
+    expect_output("quadratic_equation(9.0, -126, 441)", "(7, 7)");
+    expect_output("quadratic_equation(9, -126, 441)", "(7, 7)");
+    expect_output("quadratic_equation(1, 1, 1)", "(NaN, NaN)");
+    expect_output("quadratic_equation(1, 0, 1)", "(NaN, NaN)");
```
Comment (hamirmahal) [inline numbat/tests/interpreter.rs:235]: I'm not too sure how you'd like to handle edge cases like these.

File: numbat/src/ffi.rs [hunk 2]
```diff
@@ -724,6 +732,23 @@ fn gamma(args: &[Value]) -> Result<Value> {
     ))))
 }
 
+fn quadratic_equation(args: &[Value]) -> Result<Value> {
```
Comment (hamirmahal) [inline numbat/src/ffi.rs:735]: We can rename this to something more concise if this is a feature you want to add.

File: numbat/src/value.rs [hunk 3]
```diff
@@ -23,6 +23,7 @@ impl std::fmt::Display for FunctionReference {
 #[derive(Debug, Clone, PartialEq, Eq)]
 pub enum Value {
     Quantity(Quantity),
+    Pair(Quantity, Quantity),
```
Comment (hamirmahal) [inline numbat/src/value.rs:26]: We can go with another approach if there's a cleaner way to do this.

File: numbat/modules/math/functions.nbt [hunk 4]
```diff
@@ -36,6 +36,7 @@ fn atanh(x: Scalar) -> Scalar
 ## Others
 
 fn gamma(x: Scalar) -> Scalar
+fn quadratic_equation(a: Scalar, b: Scalar, c: Scalar) -> Scalar
```
Comment (hamirmahal) [inline numbat/modules/math/functions.nbt:39]:
```suggestion
fn quadratic_equation(a: Scalar, b: Scalar, c: Scalar) -> (Scalar, Scalar)
```
I thought this should be `(Scalar, Scalar)`, but that gave me parsing errors when trying to run tests.

Comment (mhmdanas) [reply numbat/modules/math/functions.nbt:39]: Numbat doesn't support tuples at the moment.

File: numbat/src/value.rs (hunk 3 above)
Comment (mhmdanas) [reply numbat/src/value.rs:26]: We could just return a string for now?

File: numbat/src/ffi.rs (hunk 2 above)
Comment (mhmdanas) [reply numbat/src/ffi.rs:735]: I think `solve_quadratic_equation` would be a better name.

File: numbat/src/value.rs (hunk 3 above)
Comment (hamirmahal) [reply numbat/src/value.rs]: Okay, it does that now.

File: numbat/src/ffi.rs [hunk 5]
```diff
@@ -724,6 +732,23 @@ fn gamma(args: &[Value]) -> Result<Value> {
...
+    let a = args[0].unsafe_as_quantity().unsafe_value().to_f64();
+    let b = args[1].unsafe_as_quantity().unsafe_value().to_f64();
+    let c = args[2].unsafe_as_quantity().unsafe_value().to_f64();
+
+    let root1 = (-b + f64::sqrt(b.powi(2) - 4.0 * a * c)) / (2.0 * a);
+    let root2 = (-b - f64::sqrt(b.powi(2) - 4.0 * a * c)) / (2.0 * a);
+
+    // -0 => 0
+    let root1 = if root1.abs() == 0.0 { 0.0 } else { root1 };
+    let root2 = if root2.abs() == 0.0 { 0.0 } else { root2 };
```
Comment (hamirmahal) [inline numbat/src/ffi.rs:747]: One of the tests returns `(0, -0)` without these lines.

File: book/src/list-functions.md [hunk 6]
```diff
@@ -78,6 +78,7 @@ fn acsch(x: Scalar) -> Scalar
 
 ```nbt
 fn gamma(x: Scalar) -> Scalar
+fn solve_quadratic_equation(a: Scalar, b: Scalar, c: Scalar) -> Scalar
```
Comment (mhmdanas) [inline book/src/list-functions.md:81]:
```suggestion
fn solve_quadratic_equation(a: Scalar, b: Scalar, c: Scalar) -> String
```

**Comment by hamirmahal** [comment]
> Since we're returning a string now, might as well implement the function directly in Numbat (as shown in [#397 (comment)](https://github.com/sharkdp/numbat/issues/397#issuecomment-1993827924))?

Either way is fine. Feel free to close this if this isn't something you want to add.

**Comment by mhmdanas** [comment]
I'm fine with adding it to Numbat myself (no idea about @sharkdp), it's just that I don't see a good reason to do it via FFI'ing to Rust instead of implement it directly in `numbat/modules` in Numbat code.

File: numbat/modules/math/functions.nbt [hunk 7]
```diff
@@ -37,6 +37,11 @@ fn atanh(x: Scalar) -> Scalar
 
 fn gamma(x: Scalar) -> Scalar
 
+fn _qe_solution<A, B>(a: A, b: B, c: B²/A, sign: Scalar) -> B/A =
+  (-b + sign × sqrt(b² - 4 a c)) / 2 a
+fn solve_quadratic_equation<A2, B2>(a: A2, b: B2, c: B2²/A2) -> String =
+  "x₁ = {_qe_solution(a, b, c, 1)}; x₂ = {_qe_solution(a, b, c, -1)}"
```
Comment (mhmdanas) [inline numbat/modules/math/functions.nbt:43]:
```suggestion
fn _qe_solution<A, B>(a: A, b: B, c: B²/A, sign: Scalar) -> B/A =
  (-b + sign × sqrt(b² - 4 a c)) / 2 a

fn solve_quadratic_equation<A, B>(a: A, b: B, c: B²/A) -> String =
  "x₁ = {_qe_solution(a, b, c, 1)}; x₂ = {_qe_solution(a, b, c, -1)}"
```

**Review by mhmdanas — CHANGES_REQUESTED** [review]

File: numbat/tests/interpreter.rs [hunk 8]
```diff
@@ -227,6 +227,20 @@ fn test_function_inverses() {
 
 #[test]
 fn test_math() {
+    expect_output("solve_quadratic_equation(9.0, -126, 441)", "x₁ = 7; x₂ = 7");
+    expect_output("solve_quadratic_equation(9, -126, 441)", "x₁ = 7; x₂ = 7");
+    expect_output("solve_quadratic_equation(1, 1, 1)", "x₁ = NaN; x₂ = NaN");
```
Comment (sharkdp) [inline numbat/tests/interpreter.rs:232]: This is obviously something we should implement properly. If b² - 4 a c < 0, we should show "no real-valued solution" or similar. And if b² - 4 a c == 0, we should only show the single solution -b/2a.

File: numbat/modules/math/functions.nbt [hunk 9]
```diff
@@ -37,6 +37,16 @@ fn atanh(x: Scalar) -> Scalar
 
 fn gamma(x: Scalar) -> Scalar
 
+fn _qe_solution<A, B>(a: A, b: B, c: B²/A, sign: Scalar) -> B/A =
+  (-b + sign × sqrt(b² - 4 a c)) / 2 a
+
+fn solve_quadratic_equation<A, B>(a: A, b: B, c: B²/A) -> String =
+  if b² - 4 a c < 0
```
Comment (hamirmahal) [inline numbat/modules/math/functions.nbt:44]:
```
 *  Executing task: cargo test --package numbat --test interpreter -- test_math --exact --nocapture 

    Finished test [unoptimized + debuginfo] target(s) in 0.09s
     Running tests/interpreter.rs (target/debug/deps/interpreter-ad610b883bfbc54a)

running 1 test
thread 'test_math' panicked at numbat/tests/common.rs:27:21:
called `Result::unwrap()` on an `Err` value: TypeCheckError(IncompatibleDimensions(IncompatibleDimensionsError { span_operation: Span { start: SourceCodePositition { byte: 1000, line: 44, position: 17 }, end: SourceCodePositition { byte: 1001, line: 44, position: 18 }, code_source_id: 9 }, operation: "comparison", span_expected: Span { start: SourceCodePositition { byte: 988, line: 44, position: 6 }, end: SourceCodePositition { byte: 999, line: 44, position: 16 }, code_source_id: 9 }, expected_name: " left hand side", expected_type: Product { factors: [BaseRepresentationFactor("B", Ratio { numer: 2, denom: 1 })] }, expected_dimensions: [], span_actual: Span { start: SourceCodePositition { byte: 1002, line: 44, position: 19 }, end: SourceCodePositition { byte: 1003, line: 44, position: 20 }, code_source_id: 9 }, actual_name: "right hand side", actual_name_for_fix: "expression on the right hand side", actual_type: Product { factors: [] }, actual_dimensions: ["Angle", "Scalar", "SolidAngle"] }))
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
test test_math ... FAILED

failures:

failures:
    test_math

test result: FAILED. 0 passed; 1 failed; 0 ignored; 0 measured; 30 filtered out; finished in 0.07s

error: test failed, to rerun pass `-p numbat --test interpreter`

 *  The terminal process "cargo 'test', '--package', 'numbat', '--test', 'interpreter', '--', 'test_math', '--exact', '--nocapture'" failed to launch (exit code: 101). 
 *  Terminal will be reused by tasks, press any key to close it. 
```

Comment (hamirmahal) [inline numbat/modules/math/functions.nbt:44]: It looks like this line gives an incompatible dimensions error. Maybe we can fix it by casting `0`?

Comment (hamirmahal) [reply numbat/modules/math/functions.nbt:44]: I tried `0B²` and `0 as B²`, but both gave me `UnknownIdentifier` TypeCheckErrors.

Comment (mhmdanas) [reply numbat/modules/math/functions.nbt:44]: You could sidestep the issue and use `b² < 4 a c` instead.

Same thing below for `b² - 4 a c == 0`, can be replaced with `b² == 4 a c`

**Comment by hamirmahal** [comment]
I'm happy to squash these down to one commit after another round of feedback.

File: numbat/modules/math/functions.nbt (hunk 9 above)
Comment (sharkdp) [reply numbat/modules/math/functions.nbt]: Yes, this is a limitation of the language at the moment, see #37

File: numbat/tests/interpreter.rs [hunk 10]
```diff
@@ -227,6 +227,19 @@ fn test_function_inverses() {
 
 #[test]
 fn test_math() {
+    expect_output("solve_quadratic_equation(1, 0, -1)", "x₁ = 1; x₂ = -1");
+    expect_output("solve_quadratic_equation(9.0, -126.0, 441.0)", "x = 7");
+    expect_output("solve_quadratic_equation(9.0, -126.0, 441)", "x = 7");
+    expect_output("solve_quadratic_equation(9.0, -126, 441)", "x = 7");
+    expect_output("solve_quadratic_equation(9, -126, 441)", "x = 7");
```
Comment (mhmdanas) [inline numbat/tests/interpreter.rs:234]: There's no point in having all these four lines, they're the exact same (Numbat doesn't have an integer type):

```suggestion
    expect_output("solve_quadratic_equation(9, -126, 441)", "x = 7");
```

Sorry for all the bikeshedding and nitpicking, almost there!

**Review by mhmdanas — CHANGES_REQUESTED** [review]

File: numbat/tests/interpreter.rs [hunk 11]
```diff
@@ -227,6 +227,16 @@ fn test_function_inverses() {
...
 fn test_math() {
+    expect_output("solve_quadratic_equation(1, 0, -1)", "x₁ = 1; x₂ = -1");
+    expect_output("solve_quadratic_equation(9, -126, 441)", "x = 7");
+    expect_output("solve_quadratic_equation(1, -2, 1)", "x = 1");
+    expect_output("solve_quadratic_equation(1, 0, 0)", "x = 0");
+    expect_output(
+        "solve_quadratic_equation(1, 1, 1)",
+        "no real-valued solution",
+    );
+    expect_failure("solve_quadratic_equation(0, 1, 1)", "Division by zero");
```
[the remark this answers is not in the document]
Comment (hamirmahal) [reply numbat/tests/interpreter.rs:238]: I think I run into a similar problem with `0` not being polymorphic...

[the remark this answers is not in the document]
Comment (mhmdanas) [reply numbat/tests/interpreter.rs:238]: Try `a == 0 a`?

[the remark this answers is not in the document]
Comment (mhmdanas) [reply numbat/tests/interpreter.rs:238]: Probably a good idea to also add a comment clarifying that this could be avoided by #37.

File: numbat/tests/interpreter.rs [hunk 12]
```diff
@@ -227,6 +227,18 @@ fn test_function_inverses() {
...
+    expect_output("solve_quadratic_equation(0, 9, 3)", "x = -0.333333");
+    expect_output("solve_quadratic_equation(9, -126, 441)", "x = 7");
+    expect_output("solve_quadratic_equation(1, -2, 1)", "x = 1");
+    expect_output("solve_quadratic_equation(0, 1, 1)", "x = -1");
+    expect_output("solve_quadratic_equation(1, 0, 0)", "x = 0");
+    expect_output(
+        "solve_quadratic_equation(1, 1, 1)",
+        "no real-valued solution",
+    );
+    expect_failure("solve_quadratic_equation(0, 0, 1)", "Division by zero");
```
Comment (hamirmahal) [inline numbat/tests/interpreter.rs:240]: Maybe we just want to return "no solution" here?

Comment (hamirmahal) [reply numbat/tests/interpreter.rs:240]: Wait, unless `c == 0`

File: numbat/modules/math/functions.nbt (hunk 7 above)
Comment (mhmdanas) [reply numbat/modules/math/functions.nbt]: Okay, replacing `A2` and `B2` with `A` and `B` respectively in `solve_quadratic_equation` causes this error:

```
error: while type checking                                                                                                                         
   ┌─ Module 'math::functions', File <builtin>/modules/math/functions.nbt:40:35                                                                    
   │                                
40 │ fn _qe_solution<A, B>(a: A, b: B, c: B²/A, sign: Scalar) -> B/A =   
   │                                   ^ Multiple unresolved generic parameters in a single function parameter type are not (yet) supported. Consid
er reordering the function parameters                                                                                                              
   ·                                                                                                                                               
47 │   else if b² <= 4 a c then if b² < 4 a c then "no real-valued solution" else "x = {-b / 2 a}" else "x₁ = {_qe_solution(a, b, c, 1)}; x₂ = {_qe
_solution(a, b, c, -1)}"                                                 
   │                                                                                                           ------------ In this function call  
                                                                         
Interpreter error in Prelude code 
```

@sharkdp I don't think this is supposed to happen?

Comment (mhmdanas) [reply numbat/modules/math/functions.nbt]: Ah this is https://github.com/sharkdp/numbat/issues/166

**Comment by hamirmahal** [comment]
You're welcome!

File: numbat/modules/math/functions.nbt [hunk 13]
```diff
@@ -37,6 +37,15 @@ fn atanh(x: Scalar) -> Scalar
...
 fn gamma(x: Scalar) -> Scalar
 
+fn _qe_solution<A, B>(a: A, b: B, c: B²/A, sign: Scalar) -> B/A =
+  (-b + sign × sqrt(b² - 4 a c)) / 2 a
+
+fn solve_quadratic_equation<A2, B2>(a: A2, b: B2, c: B2²/A2) -> String =
+  # TODO: Replace `a == 0 a` check after implementing https://github.com/sharkdp/numbat/issues/37.
+  if a == 0 a
+  then if b == 0 b then if c == 0 c then "infinitely many solutions" else "no solution" else "x = {-c / b}"
+  else if b² <= 4 a c then if b² < 4 a c then "no real-valued solution" else "x = {-b / 2 a}" else "x₁ = {_qe_solution(a, b, c, 1)}; x₂ = {_qe_solution(a, b, c, -1)}"
```
[the remark this answers is not in the document]
Comment (hamirmahal) [reply numbat/modules/math/functions.nbt]: Sure, how's f7d0347?

**Comment by hamirmahal** [comment]
I renamed it from `solve_quadratic_equation` to `quadratic_equation` because the solver doesn't show up as a suggestion when typing in `quad` otherwise.

image

**Comment by hamirmahal** [comment]
@sharkdp should I close this?

https://github.com/sharkdp/numbat/issues/397#issuecomment-2010396657

**Comment by sharkdp** [comment]
How about we move this to a module in `extra::` and make it "opt in"? Something like `extra::algebra`. Then I'd be fine with merging it.

**Comment by hamirmahal** [comment]
> (signal: 11, SIGSEGV: invalid memory reference)

https://github.com/sharkdp/numbat/actions/runs/8424920106/job/23069903095?pr=398#step:8:267

I'm not too sure what caused that, or how to fix it.

<details>
<summary>When I try to reproduce the issue locally I get a different error.</summary>

```
hamir@hamir-desktop:~/numbat (feat/functionality-for-quadratic-equation)$ cargo build --locked --release --target=x86_64-apple-darwin
   Compiling libc v0.2.152
   Compiling cfg-if v1.0.0
   Compiling proc-macro2 v1.0.78
   Compiling generic-array v0.14.7
   Compiling typenum v1.17.0
   Compiling siphasher v0.3.11
   Compiling num-traits v0.2.17
   Compiling phf_shared v0.11.2
   Compiling rand_core v0.6.4
   Compiling regex-syntax v0.8.2
   Compiling rand v0.8.5
   Compiling quote v1.0.35
   Compiling cc v1.0.83
   Compiling syn v2.0.48
   Compiling bitflags v2.4.2
   Compiling tinyvec_macros v0.1.1
   Compiling getrandom v0.2.12
   Compiling rustix v0.38.30
   Compiling tinyvec v1.6.0
   Compiling errno v0.3.8
   Compiling phf_generator v0.11.2
   Compiling crypto-common v0.1.6
   Compiling block-buffer v0.10.4
   Compiling spin v0.9.8
   Compiling untrusted v0.9.0
   Compiling regex-automata v0.4.4
   Compiling rustls-pki-types v1.1.0
   Compiling option-ext v0.2.0
   Compiling serde v1.0.195
   Compiling dirs-sys v0.4.1
   Compiling digest v0.10.7
   Compiling unicode-normalization v0.1.22
   Compiling ring v0.17.7
   Compiling phf_codegen v0.11.2
   Compiling phf v0.11.2
   Compiling num-integer v0.1.45
   Compiling cpufeatures v0.2.12
   Compiling log v0.4.20
   Compiling itoa v1.0.10
   Compiling unicode-bidi v0.3.15
   Compiling percent-encoding v2.3.1
   Compiling rustls v0.22.2
   Compiling utf8parse v0.2.1
   Compiling same-file v1.0.6
   Compiling sha2 v0.10.8
The following warnings were emitted during compilation:

warning: ring@0.17.7: cc: error: unrecognized debug output level ‘full’
warning: ring@0.17.7: cc: error: unrecognized command-line option ‘-arch’

error: failed to run custom build command for `ring v0.17.7`

Caused by:
  process didn't exit successfully: `/home/hamir/numbat/target/release/build/ring-393541a1edcf127a/build-script-build` (exit status: 1)
  --- stdout
  cargo:rerun-if-env-changed=RING_PREGENERATE_ASM
  cargo:rustc-env=RING_CORE_PREFIX=ring_core_0_17_7_
  OPT_LEVEL = Some("3")
  TARGET = Some("x86_64-apple-darwin")
  HOST = Some("x86_64-unknown-linux-gnu")
  cargo:rerun-if-env-changed=CC_x86_64-apple-darwin
  CC_x86_64-apple-darwin = None
  cargo:rerun-if-env-changed=CC_x86_64_apple_darwin
  CC_x86_64_apple_darwin = None
  cargo:rerun-if-env-changed=TARGET_CC
  TARGET_CC = None
  cargo:rerun-if-env-changed=CC
  CC = None
  RUSTC_LINKER = None
  cargo:rerun-if-env-changed=CROSS_COMPILE
  CROSS_COMPILE = None
  cargo:rerun-if-env-changed=CRATE_CC_NO_DEFAULTS
  CRATE_CC_NO_DEFAULTS = None
  DEBUG = Some("false")
  CARGO_CFG_TARGET_FEATURE = Some("cmpxchg16b,fxsr,sse,sse2,sse3,sse4.1,ssse3")
  cargo:rerun-if-env-changed=CFLAGS_x86_64-apple-darwin
  CFLAGS_x86_64-apple-darwin = None
  cargo:rerun-if-env-changed=CFLAGS_x86_64_apple_darwin
  CFLAGS_x86_64_apple_darwin = None
  cargo:rerun-if-env-changed=TARGET_CFLAGS
  TARGET_CFLAGS = None
  cargo:rerun-if-env-changed=CFLAGS
  CFLAGS = None
  running: "cc" "-O3" "-ffunction-sections" "-fdata-sections" "-fPIC" "-m64" "-arch" "x86_64" "-I" "include" "-I" "/home/hamir/numbat/target/x86_64-apple-darwin/release/build/ring-4e4393dd11ac0c15/out" "-Wall" "-Wextra" "-fvisibility=hidden" "-std=c1x" "-pedantic" "-Wall" "-Wextra" "-Wbad-function-cast" "-Wcast-align" "-Wcast-qual" "-Wconversion" "-Wenum-compare" "-Wfloat-equal" "-Wformat=2" "-Winline" "-Winvalid-pch" "-Wmissing-field-initializers" "-Wmissing-include-dirs" "-Wnested-externs" "-Wredundant-decls" "-Wshadow" "-Wsign-compare" "-Wsign-conversion" "-Wstrict-prototypes" "-Wundef" "-Wuninitialized" "-Wwrite-strings" "-gfull" "-DNDEBUG" "-o" "/home/hamir/numbat/target/x86_64-apple-darwin/release/build/ring-4e4393dd11ac0c15/out/crypto/curve25519/curve25519.o" "-c" "crypto/curve25519/curve25519.c"
  exit status: 1
  cargo:warning=cc: error: unrecognized debug output level ‘full’

  cargo:warning=cc: error: unrecognized command-line option ‘-arch’


  --- stderr


  error occurred: Command "cc" "-O3" "-ffunction-sections" "-fdata-sections" "-fPIC" "-m64" "-arch" "x86_64" "-I" "include" "-I" "/home/hamir/numbat/target/x86_64-apple-darwin/release/build/ring-4e4393dd11ac0c15/out" "-Wall" "-Wextra" "-fvisibility=hidden" "-std=c1x" "-pedantic" "-Wall" "-Wextra" "-Wbad-function-cast" "-Wcast-align" "-Wcast-qual" "-Wconversion" "-Wenum-compare" "-Wfloat-equal" "-Wformat=2" "-Winline" "-Winvalid-pch" "-Wmissing-field-initializers" "-Wmissing-include-dirs" "-Wnested-externs" "-Wredundant-decls" "-Wshadow" "-Wsign-compare" "-Wsign-conversion" "-Wstrict-prototypes" "-Wundef" "-Wuninitialized" "-Wwrite-strings" "-gfull" "-DNDEBUG" "-o" "/home/hamir/numbat/target/x86_64-apple-darwin/release/build/ring-4e4393dd11ac0c15/out/crypto/curve25519/curve25519.o" "-c" "crypto/curve25519/curve25519.c" with args "cc" did not execute successfully (status code exit status: 1).
```
</details>

File: book/src/list-functions.md [hunk 14]
```diff
@@ -74,6 +74,7 @@ fn sech(x: Scalar) -> Scalar
 fn asech(x: Scalar) -> Scalar
 fn csch(x: Scalar) -> Scalar
 fn acsch(x: Scalar) -> Scalar
+fn quadratic_equation<A2, B2>(a: A2, b: B2, c: B2²/A2) -> String
```
Comment (sharkdp) [inline book/src/list-functions.md:77]: Can we please move this to a new algebra section at the bottom of this page?

Comment (hamirmahal) [reply book/src/list-functions.md]: Sure, let me know if [5c7f33b](https://github.com/sharkdp/numbat/pull/398/commits/5c7f33b2b2a52b18b72d426087ca11b7cfd07436) isn't what you wanted.

**Review by sharkdp — APPROVED** [review]
Thank you

**Comment by hamirmahal** [comment]
You're welcome!

**Comment by hamirmahal** [comment]
Thank you for building this really awesome tool in Rust.

**Comment by hamirmahal** [comment]
Thank you for building this really awesome tool, and thank you for building it with Rust.

**Comment by mhmdanas** [comment]
Thanks!

**Comment by hamirmahal** [comment]
You're welcome!
